src/bubo-ht.h

```diff
@@ -166,185 +114,100 @@ class BuboHashSet {
         stat->entries = num_entries_;
-        stat->total_chain_len = 0;
-        stat->collision_slots = 0;
-        stat->max_chain_len = 0;
-        stat->avg_chain_len = 0;
-        stat->dist_1_2 = 0;
-        stat->dist_3_5 = 0;
-        stat->dist_6_9 = 0;
-        stat->dist_10_ = 0;
 
         stat->ht_bytes = table_size_ * sizeof(Entry);
 
         for (uint32_t idx = 0; idx < table_size_; idx++) {
-            Entry* p = &table_[idx];
-            uint64_t chain_len = 0;
-
-            p = p->next_;
-            if (p) stat->collision_slots ++;
-
-            while (p) {
-                stat->ht_bytes += sizeof(Entry);
-                chain_len ++;
-                p = p->next_;
-            }
-
-            if (chain_len >=1 && chain_len <=2) stat->dist_1_2 ++;
-            else if (chain_len >=3 && chain_len <=5) stat->dist_3_5 ++;
-            else if (chain_len >=6 && chain_len <=9) stat->dist_6_9 ++;
-            else if (chain_len >=10) stat->dist_10_ ++;
-
-            stat->total_chain_len += chain_len;
-            if (chain_len > stat->max_chain_len) stat->max_chain_len = chain_len;
-        }
-
-        if (stat->collision_slots > 0) {
-            stat->avg_chain_len = (double) stat->total_chain_len/ (double)stat->collision_slots;
-        } else {
-            assert(stat->total_chain_len == 0);
+            stat->ht_bytes += sizeof(Entry);
         }
 
         uint64_t allocated_bytes = 0, used_bytes = 0;
         blob_store_->stats(&allocated_bytes, &used_bytes);
         stat->blob_allocated_bytes = allocated_bytes;
         stat->blob_used_bytes = used_bytes;
+        stat->bit_set_bytes = bit_set_.size() / 8;
 
-        stat->bytes = stat->ht_bytes + stat->blob_allocated_bytes;
+        stat->bytes = stat->ht_bytes + stat->blob_allocated_bytes + stat->bit_set_bytes;
     }
 
 protected:
     struct Entry {
         const BYTE* val_;
-        Entry* next_;
-        Entry() : val_(NULL), next_(NULL) {}
-        Entry(const BYTE* v, Entry* n) : val_(v), next_(n) {}
+        Entry() : val_(NULL) {}
+        Entry(const BYTE* v, Entry* n) : val_(v) {}
     };
 
     uint32_t table_size_;
     uint32_t max_table_size_;
 
-    uint64_t table_curr_use_;
-    uint64_t table_collisions_;
     uint64_t num_entries_;
     Entry* table_;
 
     BlobStore* blob_store_;
+    boost::dynamic_bitset<> bit_set_;
 
     H hash;
     E equals;
 
     void insert_value_into_table_at_index(const BYTE* value, Entry* table, uint32_t index) {
-        if (table[index].val_ == NULL) {
-            // not found, and spine doesn't have an entry.
-            table[index].val_ = value;
-            table[index].next_ = NULL;
-            table_curr_use_ ++;
-        } else {
-            // not found, but spine has an entry. insert new item right next to spine.
-            table[index].next_ = new Entry(value, table[index].next_);
-            table_collisions_ ++;
+        while (bit_set_[index]) {
```

Comment:
This will infinitely loop if there are already `max_table_size_` entries in the table, i.e. all slots are full and you can't resize any more, so you need to handle that case. Also adding a test  of course.